blockchain/blocks/src/tipset.rs

```diff
@@ -162,65 +150,59 @@ impl Tipset {
         self.blocks
     }
     /// Returns the smallest ticket of all blocks in the tipset
-    pub fn min_ticket(&self) -> Result<Ticket, Error> {
-        if self.blocks.is_empty() {
-            return Err(Error::NoBlocks);
-        }
-        Ok(self.blocks[0].ticket().clone())
+    pub fn min_ticket(&self) -> Ticket {
+        self.first_block().ticket().clone()
     }
     /// Returns the smallest timestamp of all blocks in the tipset
-    pub fn min_timestamp(&self) -> Result<u64, Error> {
-        if self.blocks.is_empty() {
-            return Err(Error::NoBlocks);
-        }
-        let mut min = self.blocks[0].timestamp();
-        for i in 1..self.blocks.len() {
-            if self.blocks[i].timestamp() < min {
-                min = self.blocks[i].timestamp()
-            }
-        }
-        Ok(min)
+    pub fn min_timestamp(&self) -> u64 {
+        self.blocks
+            .iter()
+            .map(|block| block.timestamp())
+            .min()
+            .unwrap()
     }
     /// Returns the number of blocks in the tipset
     pub fn len(&self) -> usize {
         self.blocks.len()
     }
-    /// Returns true if no blocks present in tipset
-    pub fn is_empty(&self) -> bool {
-        self.blocks.is_empty()
-    }
     /// Returns a key for the tipset.
-    pub fn key(&self) -> &TipSetKeys {
+    pub fn key(&self) -> &TipsetKeys {
         &self.key
     }
     /// Returns slice of Cids for the current tipset
     pub fn cids(&self) -> &[Cid] {
-        &self.key.cids()
+        self.key.cids()
     }
     /// Returns the CIDs of the parents of the blocks in the tipset
-    pub fn parents(&self) -> &TipSetKeys {
-        &self.blocks[0].parents()
+    pub fn parents(&self) -> &TipsetKeys {
+        self.first_block().parents()
     }
     /// Returns the state root for the tipset parent.
     pub fn parent_state(&self) -> &Cid {
-        self.blocks[0].state_root()
+        self.first_block().state_root()
     }
     /// Returns the tipset's calculated weight
     pub fn weight(&self) -> &BigUint {
-        &self.blocks[0].weight()
+        self.first_block().weight()
     }
 }
 
-/// FullTipSet is an expanded version of the TipSet that contains all the blocks and messages
+/// FullTipset is an expanded version of the Tipset that contains all the blocks and messages
 #[derive(Debug, PartialEq, Clone)]
 pub struct FullTipset {
     blocks: Vec<Block>,
 }
 
 impl FullTipset {
-    /// constructor
-    pub fn new(blks: Vec<Block>) -> Self {
-        Self { blocks: blks }
+    /// constructor, panics when the given vector is empty
+    pub fn new(blocks: Vec<Block>) -> Self {
+        assert!(!blocks.is_empty());
```

Comment:
This isn't safe, can potentially receive a full tipset over the network and converted from a `TipsetBundle` that doesn't get validated before this constructor is called. I would just add a TODO to run the same validation on this as the `Tipset` so we can't panic on runtime here

Reply:
Ah, okay, we should probably abstract over `Tipset` and `FullTipset` then

Reply:
Exactly, this is what I was hinting at earlier. Makes sense to do in another issue since there are other changes to make along with it, unless you'd like to do now

Reply:
I'm not sure which other changes would be made alongside, there is very marginal duplication between the two types and I'm hesitant about leaking this abstraction outside of this file since you would need to `use` the trait every time

Reply:
You can do something like make it generic on an iterator instead of generic on this trait, and just map the full tipset before. I made the changes here as an example:
https://github.com/ChainSafe/forest/commit/1642d571f85825d4bcca585024b20dfb2d39dd15

Do with it how you wish, this is just one option, but I agree requiring that new trait probably isn't ergonomic

Reply:
Ah yeah that is even better